Fixed RailSwitch

Protonull opened the pull request.

Protonull committed:
Added and applied .editorconfig

Protonull committed:
Re-organised and updated the POM

Now has the most up to date dependencies for 1.12.2, and now includes Devoted's style checker

Protonull committed:
Cleaned up RailSwitch

Protonull committed:
Updated DetectorRailActivateListener

The file has been updated and improved and flattened. While some of the comments might seem a little obvious, that you can reasonably easily deduce what's going on, it's in my opinion better to save future developers, especially newer developers the effort of having to parse code. The comments allow me to define what's happening but also *why* it's happening.

Protonull committed:
Updated SetDestinationCommand

Improved, updated, and flattened the set command handler.

Protonull committed:
Oopsie

Protonull committed:
Rename RailSwitch to RailSwitchPlugin

Protonull committed:
Moved DetectorRailActivateListener into the main project folder.

If the set command doesn't have its own package, why should the listener have one? It's not like the database package where it's actually necessary to distinguish them.

Protonull committed:
Package cleanup for fcb3a76

Protonull committed:
Changed config option timings to debug

The previous option "timings" is supported for backwards compatibility.

Protonull committed:
Removed AbstractConnectionPool

It's not used by anything, not even the MySQLConnectionPool which uses all of the parameters defined.

Protonull committed:
RailSwitch now conforms to Devoted's style checker

Protonull committed:
Fixed weird XRef error

Protonull committed:
Bump version - 1.3.4

Protonull committed:
Updated plugin.yml

Protonull committed:
Not sure why this was set to 1.14, this is the 1.12 branch

Maxopoly commented on src/main/java/sh/okx/railswitch/DetectorRailActivateListener.java:
```diff
@@ -0,0 +1,163 @@
+        Block above = block.getRelative(BlockFace.UP);
+        if (above == null
+                // Do not check for Material.SIGN as that is an item material, not a block material
+                || (above.getType() != Material.SIGN_POST && above.getType() != Material.WALL_SIGN)
```
Sign posts can't be above rails

Protonull replied:
Removed

Maxopoly commented on src/main/java/sh/okx/railswitch/DetectorRailActivateListener.java:
```diff
@@ -0,0 +1,163 @@
+        if (above == null
+                // Do not check for Material.SIGN as that is an item material, not a block material
+                || (above.getType() != Material.SIGN_POST && above.getType() != Material.WALL_SIGN)
+                || !(above.getState() instanceof Sign)) {
```
Kinda redundant, but eh

Maxopoly commented on src/main/java/sh/okx/railswitch/SwitchType.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Switch type matcher, will match a type to a tag.
+ */
+public enum SwitchType {
```
This is cool

Maxopoly commented on src/main/java/sh/okx/railswitch/DetectorRailActivateListener.java:
```diff
@@ -0,0 +1,163 @@
+            return;
+        }
+        
+        // If Citadel is enabled, check that the sign and the rail are on the same group
```
I'd do the player check before the reinforcement check to waste less time on empty carts (especially hopper carts are used a ton in some farms)

Protonull replied:
Done

Maxopoly commented on src/main/java/sh/okx/railswitch/DetectorRailActivateListener.java:
```diff
@@ -0,0 +1,163 @@
+            double searchDistance = Double.MAX_VALUE;
+            for (Entity entity : block.getWorld().getNearbyEntities(block.getLocation(), 3, 3, 3)) {
+                if (entity.getType() != EntityType.PLAYER
+                        || !(entity instanceof Player)) {
```
EntityType.PLAYER guarantees it to be an instanceof Player by design

Maxopoly commented on src/main/java/sh/okx/railswitch/DetectorRailActivateListener.java:
```diff
@@ -0,0 +1,163 @@
+        // Determine whether a player has a destination that matches one of the destinations
+        // listed on the switch signs, or match if there's a wildcard.
+        boolean matched = false;
+        String[] playerDestinations = plugin.getDatabase().
```
Getting rid of this database stuff and using a CMC player setting instead would be desireable, but if you dont feel like doing that its fine.

Protonull replied:
That would be good, but that's a 1.14 CMC feature

Maxopoly commented on src/main/java/sh/okx/railswitch/DetectorRailActivateListener.java:
```diff
@@ -0,0 +1,163 @@
+            case INVERTED:
+                event.setNewCurrent(matched ? 0 : 15);
+                break;
+            // Gotta have this or the Devoted style checker will screech
```
Yes,  because type could be null

Protonull replied:
Nope :P there's a null check further up the code

Protonull replied:
Nope :P there's a null check further up the code

https://github.com/CivClassic/RailSwitch/pull/1/files#diff-4bc7691f4c6d69c86915a3d11aabf83aR70

Maxopoly commented on src/main/java/sh/okx/railswitch/SetDestinationCommand.java:
```diff
@@ -1,44 +1,80 @@
+            if (Strings.isNullOrEmpty(argument)) {
+                continue;
+            }
+            if (argument.length() > 40 || !CharMatcher.inRange('0', '9').
```
Should probably just use a single Regex here

Protonull replied:
Done

Maxopoly commented on src/main/java/sh/okx/railswitch/database/RailSwitchDatabase.java:
```diff
@@ -15,104 +13,138 @@
-            }
-          }
+    
+    private static final String DESTINATION_TABLE = "rsw_dest";
```
Not sure why this stuff is made so complicated. It's not like you'd ever want to change the table name and having the sql statements inline makes the code so much easier than this

Maxopoly commented on src/main/java/sh/okx/railswitch/database/SQLiteConnectionPool.java:
```diff
@@ -6,15 +6,26 @@
+/**
+ * Database connection implementation for SQLite databases.
+ */
 public class SQLiteConnectionPool implements ConnectionPool {
```
As far as I'm concerned we don't need to maintain this at all

Protonull replied:
Agreed, in the 1.14 version we can remove all the database stuff and rely on player settings, though we should ensure that player settings wont end up causing the same problem that NameLayer did.

Maxopoly reviewed (commented):
All my comments are just optional "nice to have". Feel free to apply whatever you feel like and ignore the rest. Did you test this?

Protonull commented:
This has been tested on my personal civ server setup and everything works and no errors are thrown

Protonull committed:
Moved the Citadel check

As advised, the Citadel check has been moved below the player check, as it is common for player's to use hopper minecarts around farms, and best to not keep querying the database.

Protonull committed:
Removed Material.SIGN_POST check

Protonull committed:
Set Destination now uses a regex

Protonull committed:
Include cleanup
